crates/containerd-shim-wasm/src/sandbox/containerd/client.rs

```diff
@@ -515,44 +539,82 @@ impl Client {
         can_precompile: bool,
         precompile_id: &String,
         needs_precompile: &mut bool,
-    ) -> std::prelude::v1::Result<WasmLayer, ShimError> {
-        let mut digest_to_load = original_config.digest().clone();
-        if can_precompile {
-            let info = self.get_info(&digest_to_load).await?;
-            if let Some(label) = info.labels.get(precompile_id) {
-                // Safe to unwrap here since we already checked for the label's existence
-                digest_to_load = label.parse()?;
-                log::info!(
-                    "layer {} has pre-compiled content: {} ",
-                    info.digest,
-                    &digest_to_load
-                );
+        media_type_label: &String,
+        all_layers: &mut HashMap<Digest, WasmLayer>,
+    ) -> std::prelude::v1::Result<(), ShimError> {
+        let parent_digest = original_config.digest().clone();
+        let digests_to_load = if can_precompile {
+            let info = self.get_info(&parent_digest).await?;
+            let child_digests = info
+                .labels
+                .into_iter()
+                .filter_map(|(key, child_digest)| {
+                    if key.starts_with(&format!("{precompile_id}/child")) {
+                        log::debug!("layer {parent_digest} has child layer: {child_digest} ");
+                        Some(child_digest)
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<_>>();
+
+            if child_digests.is_empty() {
+                vec![parent_digest.clone()]
+            } else {
+                child_digests
+                    .into_iter()
+                    .map(|d| d.parse().map_err(ShimError::Oci))
+                    .collect::<Result<Vec<Digest>>>()?
             }
-        }
-        log::debug!("loading digest: {} ", &digest_to_load);
-        let res = self
-            .read_content(&digest_to_load)
-            .await
-            .map(|module| WasmLayer {
-                config: original_config.clone(),
-                layer: module,
-            });
-
-        match res {
-            Ok(res) => Ok(res),
-            Err(err) if digest_to_load == *original_config.digest() => Err(err),
-            Err(err) => {
-                log::error!("failed to load precompiled layer: {err}");
-                log::error!("falling back to original layer and marking for recompile");
-                *needs_precompile = can_precompile; // only mark for recompile if engine is capable
-                self.read_content(original_config.digest())
-                    .await
-                    .map(|module| WasmLayer {
-                        config: original_config.clone(),
-                        layer: module,
-                    })
+        } else {
+            vec![parent_digest.clone()]
```

Comment:
nit: unnecessary clone
```suggestion
            vec![parent_digest]
```